From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maciej Patelczyk <maciej.patelczyk@intel.com>
Date: Thu, 9 Jul 2026 18:26:06 +0200
Subject: fixup! drm/xe/eudebug: Introduce EU pagefault handling
 interface

Reorganize the attention scan worker as pagefaults are yet another
asynchronous actions.

Both pagefault handler and eudebug ctrl ioctl handler stops and starts
attention worker. However they are mutually exclusive and use eu_lock
and pf_fence to protect themselves.

Stop/start attention worker from critical section so that actions are
always synchronized. The attention worker may stop using eu_lock then.

Signed-off-by: Maciej Patelczyk <maciej.patelczyk@intel.com>

(cherry-picked from commit 3bdfd62cde947afdf09fd7bb1d09ec156ef60970 drm-tip-eudebug)
Signed-off-by: Bommu Krishnaiah <krishnaiah.bommu@intel.com>

---
 drivers/gpu/drm/xe/xe_eudebug.c | 25 ++++++++++---------------
 1 file changed, 10 insertions(+), 15 deletions(-)

diff --git a/drivers/gpu/drm/xe/xe_eudebug.c b/drivers/gpu/drm/xe/xe_eudebug.c
index ce5dbcb9ba20..a20ceb4979a4 100644
--- a/drivers/gpu/drm/xe/xe_eudebug.c
+++ b/drivers/gpu/drm/xe/xe_eudebug.c
@@ -1159,12 +1159,7 @@ static int do_eu_control(struct xe_eudebug *d,
 		goto out_free;
 	}
 
-	/*
-	 * Stop attention worker to avoid race
-	 * on reading / clearing attention
-	 */
-	attention_scan_cancel(xe);
-
+	/* 1st: Wait for PF to finalize */
 	ret = -EINVAL;
 	mutex_lock(&d->eu_lock);
 	pf_fence = dma_fence_get(d->pf_fence);
@@ -1181,6 +1176,9 @@ static int do_eu_control(struct xe_eudebug *d,
 		pf_fence = dma_fence_get(d->pf_fence);
 	}
 
+	/* 2nd: wait for attention stan worker to finish and stop it */
+	attention_scan_cancel(xe);
+
 	switch (arg->cmd) {
 	case DRM_XE_EUDEBUG_EU_CONTROL_CMD_INTERRUPT_ALL:
 		/* Make sure we dont promise anything but interrupting all */
@@ -1204,7 +1202,8 @@ static int do_eu_control(struct xe_eudebug *d,
 
 	if (ret == 0)
 		seqno = atomic_long_inc_return(&d->events.seqno);
-
+	/* Resume attn worker but under lock */
+	attention_scan_flush(xe);
 	mutex_unlock(&d->eu_lock);
 
 	if (ret)
@@ -1229,7 +1228,6 @@ static int do_eu_control(struct xe_eudebug *d,
 queue_put:
 	xe_exec_queue_put(q);
 	xe_file_put(xef);
-	attention_scan_flush(xe);
 	return ret;
 }
 
@@ -1703,10 +1701,8 @@ static int send_attention_event(struct xe_eudebug *d, struct xe_exec_queue *q, i
 	write_member(struct drm_xe_eudebug_event_eu_attention, ea, lrc_handle, (u64)h_lrc);
 	write_member(struct drm_xe_eudebug_event_eu_attention, ea, bitmask_size, size);
 
-	mutex_lock(&d->eu_lock);
 	event->seqno = atomic_long_inc_return(&d->events.seqno);
 	ret = xe_gt_eu_attention_bitmap(q->gt, &ea->bitmask[0], ea->bitmask_size);
-	mutex_unlock(&d->eu_lock);
 
 	if (ret) {
 		kfree(event);
@@ -4479,8 +4475,6 @@ xe_eudebug_pagefault_create(struct xe_gt *gt, struct xe_vm *vm,
 	if (!pf)
 		goto err_put_fw;
 
-	attention_scan_cancel(gt_to_xe(gt));
-
 	mutex_lock(&d->eu_lock);
 	fence = dma_fence_get(d->pf_fence);
 
@@ -4500,6 +4494,9 @@ xe_eudebug_pagefault_create(struct xe_gt *gt, struct xe_vm *vm,
 		goto err_unlock_eu_lock;
 
 	d->pf_fence = &pf_fence->base;
+
+	attention_scan_cancel(gt_to_xe(gt));
+
 	mutex_unlock(&d->eu_lock);
 
 	INIT_LIST_HEAD(&pf->list);
@@ -4543,7 +4540,6 @@ xe_eudebug_pagefault_create(struct xe_gt *gt, struct xe_vm *vm,
 
 err_unlock_eu_lock:
 	mutex_unlock(&d->eu_lock);
-	attention_scan_flush(gt_to_xe(gt));
 	kfree(pf);
 err_put_fw:
 	xe_force_wake_put(gt_to_fw(gt), fw_ref);
@@ -4610,6 +4606,7 @@ xe_eudebug_pagefault_finalize(struct xe_eudebug_pagefault *pf, bool is_err)
 		}
 
 		d->pf_fence = NULL;
+		attention_scan_flush(gt_to_xe(gt));
 		mutex_unlock(&d->eu_lock);
 
 		xe_eudebug_put(d);
@@ -4620,8 +4617,6 @@ xe_eudebug_pagefault_finalize(struct xe_eudebug_pagefault *pf, bool is_err)
 		xe_vm_put(pf->vm);
 		kfree(pf);
 	}
-
-	attention_scan_flush(gt_to_xe(gt));
 }
 
 static int send_sync_host_event(struct xe_eudebug *d,
-- 
2.43.0

